From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bryan1029384756 <23323626+Bryan1029384756@users.noreply.github.com>
Date: Sat, 17 Jun 2023 23:46:08 -0500
Subject: [PATCH] ServerSideRecorder


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 1a79374bb771b20309fb087703bfe278d90ed71d..f13581964ac8399a1d1acf651a9856d0e6502766 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1192,7 +1192,7 @@ public abstract class PlayerList {
         //Serversiderecorder start
 
         if (player != null) {
-            ReplayRecorder recorder = ((RecorderHolder)(((ServerPlayer) player).connection)).getRecorder();
+            ReplayRecorder recorder = ((RecorderHolder)(((ServerPlayer) player).connection.connection)).getRecorder();
             if (recorder != null) {
                 recorder.onPacket(packet);
             }
