From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bryan1029384756 <23323626+Bryan1029384756@users.noreply.github.com>
Date: Fri, 30 Jun 2023 21:26:04 -0500
Subject: [PATCH] Support Region Voice Recording


diff --git a/build.gradle.kts b/build.gradle.kts
index eb8e2148a7fef15ef5870ce5ec31393810043128..61f0b30f2bcf2c6a32054e16d10179fcd0a52950 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -38,7 +38,7 @@ dependencies {
     implementation("com.fasterxml.jackson.core:jackson-annotations:2.15.2") //ServersideRecording
     implementation("com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.15.2") //ServersideRecording
     implementation ("de.maxhenkel.voicechat:voicechat-api:2.4.0")  //ServersideRecording
-    implementation ("maven.modrinth:simple-voice-chat:bukkit-2.4.11")  //ServersideRecording
+    compileOnly ("maven.modrinth:simple-voice-chat:bukkit-2.4.11")  //ServersideRecording
     runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.0")
     runtimeOnly("com.mysql:mysql-connector-j:8.0.33")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
