From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bryan1029384756 <23323626+Bryan1029384756@users.noreply.github.com>
Date: Sat, 1 Jul 2023 16:39:27 -0500
Subject: [PATCH] Support Region Voice Recording


diff --git a/src/main/java/net/minecraft/serversidereplayrecorder/recorder/RegionRecorder.java b/src/main/java/net/minecraft/serversidereplayrecorder/recorder/RegionRecorder.java
index 7312a60b1af9c66efa60a42e81256ee700c7623f..870678d345fc20cf42f06c8c20318d833a8da80c 100644
--- a/src/main/java/net/minecraft/serversidereplayrecorder/recorder/RegionRecorder.java
+++ b/src/main/java/net/minecraft/serversidereplayrecorder/recorder/RegionRecorder.java
@@ -35,6 +35,8 @@ import net.minecraft.world.level.chunk.ImposterProtoChunk;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.storage.LevelData;
+import org.bukkit.Bukkit;
+
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Paths;
@@ -219,8 +221,11 @@ public class RegionRecorder extends ReplayRecorder {
         onPacket(new ClientboundPlayerPositionPacket(viewpoint.getX() + 0.5d,viewpoint.getY(),viewpoint.getZ() + 0.5d,0f,0f, Collections.emptySet(),0));
         //ready to record changes
         if(ServerSideReplayRecorderServer.config.isVoice_recording_enabled()){
-            ServerPlayer player = new ServerPlayer(ms, ms.overworld(), FAKE_GAMEPROFILE);
-            onPacket(Utils.createCustomPacket(new SecretPacket(player, UUID.randomUUID() , Voicechat.SERVER.getServer().getPort(), Voicechat.SERVER_CONFIG)));
+            if (Bukkit.getServer().getPluginManager().isPluginEnabled("Marriage")) {
+                Voicechat mp = (Voicechat) Bukkit.getServer().getPluginManager().getPlugin("Marriage");
+                ServerPlayer player = new ServerPlayer(ms, ms.overworld(), FAKE_GAMEPROFILE);
+                onPacket(Utils.createCustomPacket(new SecretPacket(player, UUID.randomUUID() , Voicechat.SERVER.getServer().getPort(), Voicechat.SERVER_CONFIG)));
+            }
         }
     }
 
